File: server-malarkey/post-receive-stomp.pl
#!/usr/bin/perl -w

use strict;
use Cwd;
use File::Basename;
use Net::STOMP::Client;

my($stomp,$peer,$oldrev,$newrev,$refname,$subject,$body,$inline,$s,$dir);
my($fname,$fpath,$suffix);

while ( $inline=<STDIN> ) 
{
  chomp($inline);
  ($oldrev,$newrev,$refname) = split(/ /,$inline,3);

  if ($oldrev eq "0000000000000000000000000000000000000000")
  {
    $s = "\nNew branch $refname\n";
  }
  elsif ($newrev eq "0000000000000000000000000000000000000000")
  {
        $s = "\nDelete branch $refname\n";
  } 
  else
  {
    $s = qx(git log --reverse -r $oldrev..$newrev);
  }

  $dir = cwd();
  ($fname,$fpath,$suffix) = fileparse($dir,'\.git');

  $subject = "[$fname] $refname $newrev";
  $body = "repo: $fname\noldrev: $oldrev\nnewrev: $newrev\nrefname: $refname\n $s\n";

  $stomp = Net::STOMP::Client->new(uri => "stomp://broker01:61613");

  $peer = $stomp->peer();

  $stomp->connect( login => "gituser", passcode => "gitpw");

  $stomp->send( destination => "/topic/future.git.commits", subject => $subject, time => time(), body => $body);
  $stomp->send( destination => "/topic/future.events.gitolite", subject => "Talking to eventbot", time => time(), body => $subject);

  $stomp->disconnect();
}

#if (open(OUTFILE,">>/var/log/stomp-rx-git.log"))
#{
# print OUTFILE "$subject\n$body\n";
#}
#close(OUTFILE);

File: server-malarkey/post-receive-stomp.rb
#!/usr/bin/env ruby

require 'rubygems'
require 'grit'
require 'yaml'
require 'stomp'
require 'syslog'

yconfig = YAML.load_file("/etc/stomp-postrec/stomp-postrec.yaml")

Syslog.open('stomp-postrec', Syslog::LOG_CONS, Syslog::LOG_DAEMON)

stompconnector = yconfig["rserver"] + "?" + yconfig["stomp-options"]
trigger_topic = yconfig["trigger-topic"]
report_topic = yconfig["report-topic"]

cdiff = " "

while msg = gets
  msg.chomp!
  old_sha, new_sha, ref = msg.split(' ', 3)

  # Syslog.info("Called with #{old_sha} / #{new_sha} / #{ref}")

  rdir = Dir.pwd
  fname = File.basename(rdir,".git")
  repo = Grit::Repo.new(rdir)

  # Syslog.info("Working dir #{rdir} / repo name #{fname}")

  if repo
    commit = repo.commit(new_sha)
    sref = ref.chomp

    subject = "[#{fname}] #{sref} #{new_sha}"

    cdiff << "#{new_sha}\nAuthor: #{commit.author}\nDate: #{commit.committed_date}\n\n#{commit.message}\n"

    body = "repo: #{fname}\noldrev: #{old_sha}\nnewrev: #{new_sha}\nrefname: #{sref}\n#{cdiff}\n"

    client = Stomp::Client.new(stompconnector)
    if client
      Syslog.info(subject)
      client.publish("/topic/#{trigger_topic}",body,{:subject => subject})
      client.publish("/topic/#{report_topic}",subject, {:subject => "Talking to eventbot"})
      client.close
    end
  else
    Syslog.info("Can't create grit instance for repo #{fname}")
  end
end
Syslog.close
